realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt

```diff
@@ -0,0 +1,279 @@
+            val doc2 = Document("hello", "world2")
+            val doc3 = Document("hello", "world3")
+
+            // Test findOne() on empty collection with no filter and no options
+            assertNull(findOne().blockingGetResult())
+
+            // Insert a document into the collection
+            insertOne(doc1).blockingGetResult()
+            assertEquals(1, count().blockingGetResult())
+
+            // Test findOne() with no filter and no options
+            assertEquals(doc1, findOne().blockingGetResult()!!.withoutId())
+
+            // Test findOne() with filter that does not match any documents and no options
+            assertNull(findOne(Document("hello", "worldDNE")).blockingGetResult())
+
+//            // Insert 2 more documents into the collection
+////            insertMany(listOf(doc2, doc3)).blockingGetResult()    // use insertOne for now
+//            insertOne(doc2).blockingGetResult()
+//            insertOne(doc3).blockingGetResult()
+//            assertEquals(3, count().blockingGetResult())
+//
+//            // test findOne() with projection and sort options
+//            val projection = Document("hello", 1)
+//            projection["_id"] = 0
+//            val options1 = RemoteFindOptions()
+//                    .limit(2)
+//                    .projection(projection)
+//                    .sort(Document("hello", 1))
+//            assertEquals(doc1, findOne(Document(), options1).blockingGetResult()!!.withoutId())
+//
+//            val options2 = RemoteFindOptions()
+//                    .limit(2)
+//                    .projection(projection)
+//                    .sort(Document("hello", -1))
+//            assertEquals(doc3.withoutId(), findOne(Document(), options2).blockingGetResult()!!.withoutId())
+//
+//            // test findOne() properly fails
+//            try {
+//                Tasks.await(coll.findOne(Document("\$who", 1)))
+//                Assert.fail()
+//            } catch (ex: ExecutionException) {
+//                Assert.assertTrue(ex.cause is StitchServiceException)
+//                val svcEx = ex.cause as StitchServiceException
+//                assertEquals(StitchServiceErrorCode.MONGODB_ERROR, svcEx.errorCode)
+//            }
+        }
+    }
+
+    // FIXME: more to come
+
+    private fun getCollectionInternal(collectionName: String, javaClass: Class<Document>? = null): MongoCollection<Document> {
+        return when (javaClass) {
+            null -> database.getCollection(collectionName)
+            else -> database.getCollection(collectionName, javaClass)
+        }
+    }
+
+    private fun Document.withId(objectId: ObjectId? = null): Document {
+        return apply { this["_id"] = objectId ?: ObjectId() }
+    }
+
+    private fun Document.withoutId(): Document {
+        return apply { remove("_id") }
+    }
+
+    private companion object {
+        const val SERVICE_NAME = "BackingDB"    // it comes from the test server's BackingDB/config.json
```

Comment:
Not entirely sure what is best practise here, but should these just be defined at the top of the file instead of in the companion object. I have a feeling that we are just putting them in Companion objects because that would make them static like in Java, but it might not be Kotlin best practise?

Reply:
Some say that if the scope of the constants is constrained to this class, it should be a companion object, whereas if its scope is wider, then it should be a top-of-the-file definition.

Reply:
You can still make it private in the top of the file, which limits the scope more or less to the file.

Reply:
That's right too. I don't know either which is considered the standard though.

Reply:
I will put them as top-level definitions. It always is nice to try something new :-)